Excited groups will care about gas reactions before dismantling. Turfs will not sleep when there are gas reactions. Fixes water vapour not consuming water vapour when freezing turfs.

Pickle-Coding opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Excited groups will no longer dismantle when there is a gas reaction occurring, and turfs no longer sleep when there's a gas reaction. Fixes water vapour not consuming water vapour when freezing turfs.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
To prevent odd phenomena where turfs with water vapour sleep, then stop applying the slippery effect despite being in conditions to react.
<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
fix: Fixes water vapour not consuming water vapour when freezing turfs.
code: Turfs will not sleep during an ongoing gas reaction, and excited groups won't dismantle if one of their turfs are reacting.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Pickle-Coding committed:
Excited groups care about reactions.

Adds reacting var for open turfs. Reacting gets set to the return of react() inside the open turf's process_cell() proc.

Adds has_reactions var for excited groups. Has_reactions gets set to the result of has_reactions OR reacting. Excited group processing will check if the excited group's has_reactions is REACTING or STOP_REACTIONS before dismantling the excited group. Excited group processing will set the excited group's has_reactions to NONE at the end before MC_TICK_CHECK.

Fixes water vapour not consuming water vapour when temperature is below WATER_VAPOR_DEPOSITION_POINT.

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -41,6 +41,8 @@
 
 	///If there is an active hotspot on us store a reference to it here
 	var/obj/effect/hotspot/active_hotspot
+	///If there is a reaction, store the result.
+	var/reacting
```
why is this a variable on /turf?

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -345,10 +347,12 @@
 
 	update_visuals()
-	if(!consider_superconductivity(starting = TRUE) && !active_hotspot) //Might need to include the return of react() here
+	if(!consider_superconductivity(starting = TRUE) && !active_hotspot && !reacting) //Might need to include the return of react() here
```
remove my old comment, also if you're gonna check the flags in excited group handling do it here too

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -407,6 +411,8 @@
 	var/display_id = 0
 	///Wrapping loop of the index colors
 	var/static/wrapping_id = 0
+	///If there are turfs in the group with gasmixtures with reactions occuring.
+	var/has_reactions = NONE
```
call this turf_reactions, and describe it as all turf reaction flags we've received. what you've got now is a bit unclear

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -109,8 +109,9 @@
 	switch(air.temperature)
 		if(-INFINITY to WATER_VAPOR_DEPOSITION_POINT)
 			if(location?.freeze_turf())
```
a: does this mean if we can't freeze the turf we won't react? What does that mean?

Pickle-Coding replied:
The reaction isn't really doing anything if that check doesn't pass, so I wouldn't really consider that as reacting.

LemonInTheDark replied:
oh TRUE then we don't set the flag
yeah nevermind this is safe

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -109,8 +109,9 @@
 			if(location?.freeze_turf())
-				SET_REACTION_RESULTS(0)
-			. = REACTING
+				air.gases[/datum/gas/water_vapor][MOLES] -= MOLES_GAS_VISIBLE
+				SET_REACTION_RESULTS(MOLES_GAS_VISIBLE)
+				. = REACTING
```
B: this is duped from below right? why not pull it out?

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -345,10 +347,12 @@
-	our_air.react(src)
+	reacting = our_air.react(src)
+	if(our_excited_group)
+		our_excited_group.has_reactions |= reacting
```
make a comment describing why we do this, somewhere

Pickle-Coding committed:
Requested changes.

Changes water vapour reaction to have a consumed var, it gets set to MOLES_GAS_VISIBLE when it freezes a turf or makes it slippery. If consumed, then water vapour moles decrease by consumed, set reaction results and set return value to REACTING.

Changes the reacting var from open turfs to be a local var for process_cell open turf proc.

Changes all occurrences of "has_reactions" to "turf_reactions".

Changes the check in process_cell after update_visuals() to check if reacting is REACTING or STOP_REACTIONS.

Pickle-Coding committed:
Adds a comment.

Adds a comment to line 350 in LINDA_turf_tile.dm to explain why we are adding this to turf_reactions.

Pickle-Coding committed:
Oops.

Fixes water_vapour reaction by making its mole count decrease by consumed instead of attempting to do arithmetic on the water vapour list, which somehow didn't runtime.

LemonInTheDark commented on code/modules/atmospherics/environmental/LINDA_turf_tile.dm:
```diff
@@ -345,10 +345,12 @@
 
 	update_visuals()
-	if(!consider_superconductivity(starting = TRUE) && !active_hotspot) //Might need to include the return of react() here
+	if(!consider_superconductivity(starting = TRUE) && !active_hotspot && !(reacting & (REACTING | STOP_REACTIONS)))
```
if you wanna be cool make a define out of  REACTING | STOP_REACTIONS that explains more of what it actually represents
